nexus/src/app/instance.rs

```diff
@@ -173,12 +173,58 @@ impl super::Nexus {
         // TODO-robustness We need to figure out what to do with Destroyed
         // instances?  Presumably we need to clean them up at some point, but
         // not right away so that callers can see that they've been destroyed.
-        let (.., authz_instance) = LookupPath::new(opctx, &self.db_datastore)
-            .organization_name(organization_name)
-            .project_name(project_name)
-            .instance_name(instance_name)
-            .lookup_for(authz::Action::Delete)
+        let (.., authz_instance, db_instance) =
+            LookupPath::new(opctx, &self.db_datastore)
+                .organization_name(organization_name)
+                .project_name(project_name)
+                .instance_name(instance_name)
+                .fetch()
+                .await?;
+
+        opctx.authorize(authz::Action::Delete, &authz_instance).await?;
+
+        match db_instance.runtime_state.state.state() {
+            InstanceState::Stopped | InstanceState::Failed => {
+                // ok
+            }
+
+            state => {
+                return Err(Error::InvalidRequest {
+                    message: format!(
+                        "instance cannot be deleted in state \"{}\"",
+                        state,
+                    ),
+                });
+            }
+        }
+
+        // Detach all attached disks
+        let disks = self
+            .instance_list_disks(
```

Comment:
I think this may be subject to a race condition: disks could be attached between when they're listed and when they're attached, which would leave those new attachments wedged. Fixing that would require more invasive surgery, and I don't think it should block this PR from merging. It might be worth a `TODO`, though.